Rework how we handle PSPs

mook-as opened the pull request:
We now allow the role manifest to list the PSPs / roles / cluster roles explicitly.  This way the deployment can customize what sort of (cluster) roles it wants (e.g. perhaps only one role needs to read node labels).

This should be backwards compatible: this should work with an unmodified SCF deployment.

mook-as committed:
Kube: Support shipping pod security policies

This adds a few related things:
- cluster roles can be defined in the role manifest
- cluster roles can be bound to accounts in the role manifest
- pod security policies can be declared in the role manifest

Note that we currently do not validate the pod security policies in any
way; we rely on Kubernetes rejecting it when we eventually deploy
things.

The old-style assumed PSPs ("privileged" and "nonprivileged") should
work for a while during transition.

Additionally, we:

- Rework the {cluster,} role {binding,} resource logic to make sure we
  include them with the instance group definitions when they're only
  used by one instance group.
- Ensure that we generate files that include accounts, then roles /
  cluster roles, and the bindings that used them, then the pods that
  use those accounts.  This makes sure `kubectl apply` doesn't get
  confused and skip creating things that don't validate (because their
  dependencies don't exist yet).

Pending work:

- Inline PSPs into instance group definitions as appropriate
- Guard against clusters without PSP support

mook-as committed:
Kube: only do the RBAC stuff if the cluster appears to have RBAC

Note that this doesn't test that RBAC is _enabled_, just that it is
supported (i.e. the cluster can handle RBAC objects).  Actually checking
that RBAC is enabled is still the task of the .Values.kube.auth knob.

jandubois reviewed (dismissed):
This PR reverts a local patch we have for `mholt/archiver`. Either that change must be excluded from the PR, or #416 must be fixed first.

mook-as commented:
Sorry about that; `dep` shenanigans (I added a dependency, then removed it again, leading to the kind of pointless vendor change).

Rebased to drop that commit (because it's not material to the PR).

andreas-kupries commented on app/fissile_test.go:
```diff
@@ -417,7 +406,7 @@ func TestGenerateAuth(t *testing.T) {
 		`auth/auth-role-pointless.yaml`: []string{
 			`{
-				"apiVersion": "rbac.authorization.k8s.io/v1beta1",
+				"apiVersion": "rbac.authorization.k8s.io/v1",
```
Does this imply a minimum version of kube will be needed to handle the output ?
If yes, are we documenting this version somewhere ?

mook-as replied:
We're already using `rbac.authorization.k8s.io/v1` for [`ClusterRoleBinding`](https://github.com/cloudfoundry-incubator/fissile/blob/c617cf6372fe75ebe255f2920c1dff6d6d0059ba/kube/rbac.go#L41), so this isn't actually a new requirement.  It's supported at least as far back as [Kubernetes 1.9](https://v1-9.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.9/#clusterrolebindinglist-v1-rbac) (and docs for older versions are busted).

andreas-kupries replied:
ok

andreas-kupries commented:
Saw the following output when building things with the new fissile for scf, i.e. within PR https://github.com/SUSE/scf/pull/1967#pullrequestreview-191781887

```
Found 8 instance groups requiring escalation: [blobstore log-api credhub-user autoscaler-api autoscaler-servicebroker cf-usb-group routing-api api-group]
make/bosh-images
Found 8 instance groups requiring escalation: [cf-usb-group routing-api api-group blobstore log-api credhub-user autoscaler-api autoscaler-servicebroker]
```

Should I see this warning ?
And two, note the changing order of entries in the list.
Should maybe have a `sort` somewhere.
I guess this comes from a map, and Go helpfully randomizes the returned order for us.

andreas-kupries commented on app/fissile.go:
```diff
@@ -791,77 +791,95 @@ func (f *Fissile) generateAuth(settings kube.ExportSettings) error {
 		}
-		outputPath := filepath.Join(authDir, fmt.Sprintf("account-%s.yaml", accountName))
-		outputFile, err := os.Create(outputPath)
+		node.Set(helm.Comment(fmt.Sprintf(`Cluster role %s used by accounts %s`, roleName, strings.Join(accountNames, "\n"))))
```
This comes out as

```
# Cluster role privileged used by accounts default-privileged
# garden-runc
# privileged
```

I would like it better to look like

```
# Cluster role "privileged" used by accounts
# - default-privileged
# - garden-runc
# - privileged
```

I.e.: Quotes around CR name, and the list of accounts looking more like a list, with "-" prefix, and all accounts on their own line.

See also line 832 for the same for roles and accounts.

This would also apparently be more consistent with what I see for service accounts, although that uses two spaces indentation, and should also have quotes around the SA name.

```
# Service account privileged is used by the following instance groups:
#  - nfs-broker
#  - router
```

Generally all the header comments do not seem to use (double)quotes to separate object names from the text around them. IMHO it would be more readable with (double)quotes.

mook-as replied:
Fixed in 12d0ca53.

andreas-kupries commented on app/fissile.go:
```diff
@@ -791,77 +791,95 @@ func (f *Fissile) generateAuth(settings kube.ExportSettings) error {
 		if err != nil {
 			return err
 		}
+		node.Set(helm.Comment(fmt.Sprintf(`Role %s used by accounts:\n%s`, roleName, strings.Join(accountNames, "\n"))))
```
Quoting issue: The `backtics` mean that the `\n` is printed as-is, instead of as newline. Use double-quotes instead. See also the comment later for related thoughts.

mook-as replied:
Fixed in 12d0ca53.

mook-as committed:
Kube: clean up formatting of resource comments

Try to be consistent about quoting resource names, except where they are
in a bullet point list, in which case don't provide an extra indent and
ensure they are sorted.

mook-as commented:
The requiring escalation warning has also been sorted in 12d0ca53. (Unfortunately, we don't currently have a way of returning warnings, just errors)

andreas-kupries reviewed (approved):
Looks good to me now. Output too.
